velox/functions/sparksql/aggregates/BloomFilterAggAggregate.cpp

```diff
@@ -0,0 +1,313 @@
+    rows.applyToSelected([&](vector_size_t row) {
+      VELOX_USER_CHECK(
+          !decodedRaw_.isNullAt(row),
+          "First argument of bloom_filter_agg cannot be null");
+      auto group = groups[row];
+      auto tracker = trackRowSize(group);
+      auto accumulator = value<BloomFilterAccumulator>(group);
+      accumulator->init(capacity_);
+      accumulator->insert(decodedRaw_.valueAt<int64_t>(row));
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    decodedIntermediate_.decode(*args[0], rows);
+    rows.applyToSelected([&](auto row) {
+      if (UNLIKELY(decodedIntermediate_.isNullAt(row))) {
+        return;
+      }
+      auto group = groups[row];
+      auto tracker = trackRowSize(group);
+      auto serialized = decodedIntermediate_.valueAt<StringView>(row);
+      auto accumulator = value<BloomFilterAccumulator>(group);
+      accumulator->mergeWith(serialized);
+    });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    decodeArguments(rows, args);
+    auto tracker = trackRowSize(group);
+    auto accumulator = value<BloomFilterAccumulator>(group);
+    accumulator->init(capacity_);
+    if (decodedRaw_.isConstantMapping()) {
+      VELOX_USER_CHECK(
+          !decodedRaw_.isNullAt(0),
+          "First argument of bloom_filter_agg cannot be null");
+      // All values are same, just do for the first.
+      accumulator->insert(decodedRaw_.valueAt<int64_t>(0));
+      return;
+    }
+    rows.applyToSelected([&](vector_size_t row) {
+      VELOX_USER_CHECK(
+          !decodedRaw_.isNullAt(row),
+          "First argument of bloom_filter_agg cannot be null");
+      accumulator->insert(decodedRaw_.valueAt<int64_t>(row));
+    });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    decodedIntermediate_.decode(*args[0], rows);
+    auto tracker = trackRowSize(group);
+    auto accumulator = value<BloomFilterAccumulator>(group);
+    rows.applyToSelected([&](auto row) {
+      if (UNLIKELY(decodedIntermediate_.isNullAt(row))) {
+        return;
+      }
+      auto serialized = decodedIntermediate_.valueAt<StringView>(row);
+      accumulator->mergeWith(serialized);
+    });
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    VELOX_CHECK(result);
+    auto flatResult = (*result)->asUnchecked<FlatVector<StringView>>();
+    flatResult->resize(numGroups);
+
+    int32_t totalSize = getTotalSize(groups, numGroups);
+    Buffer* buffer = flatResult->getBufferWithSpace(totalSize);
```

Comment:
This works, but is quite a bit of code and easy to get wrong. For example, by forgetting to call buffer->setSize or forgetting to add buffer_.size() when initializing bufferPtr.

Consider introducing new method:

```
char* rawBuffer = flatResult->getRawStringBufferWithSpace(totalSize);
```

This method would return the pointer to the first "writable" byte and update the size of the 'buffer' to include totalSize.